designs/aws/custom-user-data-and-amis.md

```diff
@@ -0,0 +1,285 @@
+2. It forces users to continue to manage their infrastructure via some tooling like Terraform or CDK, and prevents them from going kube-native for everything except their EKS cluster.
+3. Node properties that Karpenter should own (labels, taints, maxPods, nodeAllocatable) are now dictated by the user. This can lead to surprises such as nodes being orphaned because of missing provisioner labels, inaccurate binpacking leading to evictions and so on.
+
+Our current API is versioned and we’re on `v1alpha5`. This gives us the flexibility to redefine the UX to better suit everyone, even if it were to be backwards incompatible with our GA experience.
+
+
+Before we determine how the Provisioner API needs to change, we need to first agree on what the final user experience should be. As part of the [first draft](https://github.com/aws/karpenter/pull/1270) of the launch templates UX redefinition, we’ve already decided that we will not support partial launch templates, not express instance configuration requirements via the podSpec, and rather focus on strongly typing any required instance configuration within the provisionerSpec itself. This document now attempts to further dive into what the UserData, AMI and other EC2 property inline support will look like.
+
+------
+------
+
+
+### Supporting UserData
+
+We have 16* upvotes for [this request](https://github.com/aws/karpenter/issues/885) currently and several more reachouts on Slack. Most userData customizations generally fall into the following two buckets:
+
+* _CASE 1_ - Installing some binaries on worker nodes. These are generally custom tooling, security daemons etc.
+* _CASE 2_ - Modifying the kubelet, container-runtime configuration or some system level property like ulimits.
+
+
+Once we give users the ability to override the UserData, we need to determine how to merge their UserData contents with the contents that Karpenter wants to set. Look at this section in the [appendix](#appendix) for more information on why *it’s critical* that Karpenter always has control over the UserData and therefore merging is necessary.
+
+The merging logic for UserData for AL2 AMIs and BottleRocket AMIs is different. This is because the former uses cloud-init that executes UserData and the latter uses TOML which works in a different way.
+
+**For BottleRocket AMFamily**
+
+BottleRocket AMIs rely on UserData being defined [as TOML](https://github.com/toml-lang/toml) which is a markup language that is a set of unique key-value pairs. Since TOML is essentially a large dictionary / hashTable, we can accept TOML and merge contents with whatever Karpenter wants to add in. For example, Karpenter will add whatever label  and taints it thinks are accurate to ensure that there is no invalid scheduling performed. Consider the following example-
+
+A user's BR Data -
+```
```

Comment:
It's probably a good idea to specify `toml` in here so that color coding can be done correctly.